### DISCUSSION TITLE
Owner and permissions for docker-mailserver

### DISCUSSION THREAD

**Discussion by vikasrawal** [docker-mailserver/docker-mailserver#3114] [Q&A]
This seems very basic but I could not find a clear answer.  I am trying to setup DMS on a Debian server. I have kept it at /opt/docker-mailserver. Is that correct? Should the files be owned by root? Should the docker-mailserver be run as root? What permissions should different directories have (config, certs, mail-data, docker-compose.yml, mailserver.env etc)? Does DMS automatically take care of these things?

thanks for your help

**Comment by polarathene** [comment]
> I have kept it at /opt/docker-mailserver. Is that correct?

You can use any location that suits you.

---

> Should the files be owned by root? Should the docker-mailserver be run as root?

Typically Docker daemon is run as root, and thus the folders you provide as volume bind mounts will store content that is written from the containers users / groups. These aren't likely to align with your Docker host users/groups except for root.

The start-up scripts that run in the container will do their best to manage this for you, we have a step that runs `chown` on files and folders that may be provided from volumes, so that if the UID and GID values have changed (may happen between releases), they should be adjusted accordingly for you.

A volume providing a mount to `/var/mail` in the container will typically be `5000:5000` for ownership instead of root. This is set via `chown` if needed by the containers root user during container start-up, separate from the other changes handled for `/var/mail-state` volume, mentioned earlier.

Some users run DMS with a rootless Docker daemon. This is documented via contributions, but not something we officially support.

---

> What permissions should different directories have (config, certs, mail-data, docker-compose.yml, mailserver.env etc)?

The root user in the container should be able to read all files. The certs may be copied internally or used directly depending on the `SSL_TYPE` configured. You shouldn't need to worry about that though :+1: 

`docker-compose.yml` and `mailserver.env` are not used within the container. If you're new-ish to Docker, these are used to configure the container and run an instance of it. You shouldn't have any issues there, so long as the command to run a container works fine, otherwise adjust permissions accordingly.
- If you're running this on a desktop locally for example to try DMS, usually your logged in user would be part of the `docker` group so it has permission to run Docker as root, otherwise you'd need to use `sudo`.
- If deploying on a server, you can use the root user directly, or whatever makes sense to you if you have a more restricted server.

> Does DMS automatically take care of these things?

DMS is a project focused on integration of various mail software to simplify setup for users. The bulk of it is the start-up scripts, with some config files and ENV to setup a mail-server. If you're new to running a mail-server, we try to provide good defaults, but you'll still need to configure DNS, provision your certificates, etc. Our docs could do with some love, but should mostly guide you with anything you need to know, be sure to check the FAQ too.

**Reply by vikasrawal** [reply]
Thanks very much. This is very helpful. One concern I have is that many of the files including in mail-data (which is owned by 5000:5000) are readable to all users on the host machine. Should I do something like 'chmod o-rwx *' on them? Would that create a problem?

**Reply by polarathene** [reply]
> Would that create a problem?

As the start-up scripts will modify ownership each time, any change you do on the host related to that would get reverted the next time the container needs to start.

Here is the two scripts performing these changes that I am aware of:

https://github.com/docker-mailserver/docker-mailserver/blob/ae05e6a7c379825fb464a648acc04b4074dd547f/target/scripts/startup/misc-stack.sh

https://github.com/docker-mailserver/docker-mailserver/blob/ae05e6a7c379825fb464a648acc04b4074dd547f/target/scripts/helpers/utils.sh#L22-L36

They are only doing `chown` / `chgrp`, so I suppose you could run `chmod o-rwx *` on the volume data. AFAIK, modifying permissions for users not valid for assigned UID or GID of the files should be fine :sweat_smile: 

It's possible that something internal may not be happy if we've not accounted for that. None of the original maintainers or project author work on the project anymore. The current maintainers keep the project updated and running, fix bugs that are reported and review / merge features, etc.

If you do try it and find any problems occur within the container, do let us know :)
